Add axes_2d gizmo.

lambertsbennett opened the pull request:
# Objective

This PR addresses #12222 (Fixes #12222). Simple addition to add a 2D axes gizmo. 

## Solution

- Add a new method axes_2d which takes a transform and a case length and then draws two arrows in the XY plane. 

The only thing I'm not sure about here is taking a 3D transform as an argument. It says in the transform comments that for 2D the z-axis is used for ordering, so I figured I'd keep it that way?

---

## Changelog

- Add method axes_2d.
- Update arrow_2d to also calculate the tip length depending on arrow length as in arrow.
- Add axes_2d to examples 2d_gizmos.

lambertsbennett commented:
I see that the build/CI are failing because the compiler says that From<Vec3> is not implemented for Vec2. This is puzzling to me because I did implement it as:

```rust
impl From<Vec3> for Vec2 {
    #[inline]
    fn from(v: Vec3) -> Self {
        (v.x, v.y)
    }
}
```
I am relatively new to rust though, so perhaps this isn't correct? Any pointers here would be great! The strange thing is that the example works just fine (```2d_gizmos```). It's just when compiling the crate that there is an error.

mockersf commented on crates/bevy_gizmos/Cargo.toml:
```diff
@@ -14,8 +14,8 @@ webgpu = []
 # Bevy
-bevy_pbr = { path = "../bevy_pbr", version = "0.14.0-dev", optional = true }
 bevy_sprite = { path = "../bevy_sprite", version = "0.14.0-dev", optional = true }
+bevy_pbr = { path = "../bevy_pbr", version = "0.14.0-dev", optional = true }
```
you should avoid changing this

mockersf commented on examples/gizmos/2d_gizmos.rs:
```diff
@@ -87,6 +87,8 @@ fn draw_example_collection(
         YELLOW,
     );
+
+    gizmos.axes_2d(Transform::from_translation(Vec3::new(sin, 0.0, 0.0)), 25.0)
```
this example is already quite crowded. this will also put it on the grid gizmo already present, making it hard to read

mockersf commented on crates/bevy_gizmos/src/arrows.rs:
```diff
@@ -163,3 +170,36 @@ impl<'w, 's, T: GizmoConfigGroup> Gizmos<'w, 's, T> {
+        let end_x = transform.transform_point(base_length * Vec3::X);
+        let end_y = transform.transform_point(base_length * Vec3::Y);
+
+        self.arrow_2d(start.into(), end_x.into(), RED);
+        self.arrow_2d(start.into(), end_y.into(), GREEN);
```
```suggestion
        self.arrow_2d(start.xy(), end_x.xy(), RED);
        self.arrow_2d(start.xy(), end_y.xy(), GREEN);
```

and you'll need to import ``bevy_math::Vec3Swizzles`

mockersf commented on crates/bevy_gizmos/src/arrows.rs:
```diff
@@ -125,7 +125,14 @@ impl<'w, 's, T: GizmoConfigGroup> Gizmos<'w, 's, T> {
         color: impl Into<Color>,
     ) -> ArrowBuilder<'_, 'w, 's, T> {
-        self.arrow(start.extend(0.), end.extend(0.), color)
+        let length = (end - start).length();
+        ArrowBuilder {
+            gizmos: self,
+            start: start.extend(0.),
+            end: end.extend(0.),
+            color: color.into(),
+            tip_length: length / 10.,
+        }
```
Why did you change this?

lambertsbennett replied:
I changed it to be consistent with the 3d arrow drawing method, but I can remove that.

lambertsbennett commented:
Ok thanks for the feedback @notmd and @mockersf - I reverted a bunch of things and now it seems to be in a better state.

pablo-lua commented on crates/bevy_gizmos/src/arrows.rs:
```diff
@@ -163,3 +163,36 @@ impl<'w, 's, T: GizmoConfigGroup> Gizmos<'w, 's, T> {
     }
 }
+
+impl<'w, 's, T: GizmoConfigGroup> Gizmos<'w, 's, T> {
```
I think we should adress both methods in the same impl if possible.

pablo-lua reviewed (approved):
Good change, thank you. Can you put this feature in the axes example? I think we can place it in the plane on the example? We can make the bevy logo bounce around in the floor I believe.

lambertsbennett commented:
@pablo-lua I moved it into the same impl block as you suggested and added a 2D axis to the axes example. My first try at that caused there to be axes flying around in space and so I pinned them to the plane. Maybe this is not exactly illustrative of the capability of the 2D axes to transform though.

pablo-lua commented:
Fair, I was thinking in placing this as a `Sprite` in reality, like, spawning a Sprite of some kind and just use the transform in it to move it around and so. We can address this in another PR though, so I think we should revert it for now and do this as a follow up.

lambertsbennett commented:
@pablo-lua ok reverted the example, but otherwise I hope its all good!
